### ISSUE TITLE
Color by CDS might be overridden in tracks using custom CSS

### ISSUE THREAD

**Issue by childers** [GMOD/Apollo#2360]
I'm trying to set up multi-feature types in one track and have almost gotten is all working.

It looks like in my instance the custom CSS overrides the color by CDS option found in Apollo.  
Changes I've made include pulling my custom CSS out of a plugin and into the custom.css located in the jbrowse data directory. 
After some discussion with @cmdcolin I opted to not use the ```!important ``` flag in my css, to not force overrides.  

Is it possible that the CSS in my custom.css is completely overriding the color-by-CDS feature? 

Probably related to #2203 

Sample track Stanza I'm using:
``` 
      {
         "category" : "NCBI Annotation Release 100",
         "compress" : 0,
         "key" : "NCBI_Annotation_Release_100_Gene",
         "label" : "NCBI_Annotation_Release_100_Gene",
         "metadata" : {
            "Data description" : "ftp://ftp.ncbi.nlm.nih.gov/genomes/Varroa_destructor/README_CURRENT_RELEASE",
            "Data provider" : "NCBI",
            "Data source" : "ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/443/255/GCF_002443255.1_Vdes_3.0/GCF_002443255.1_Vdes_3.0_genomic.gff.gz",
            "Method" : "http://www.ncbi.nlm.nih.gov/genome/annotation_euk/process/"
         },
         "storeClass" : "JBrowse/Store/SeqFeature/NCList",
         "style" : {
            "arrowheadClass" : "trellis-arrowhead",
            "className" : "container-16px",
            "description" : "product, note, description",
            "subfeatureClasses" : {
               "CDS" : "gnomon_CDS",
               "UTR" : "gnomon_UTR",
               "exon" : "container-100pct",
               "wholeCDS" : null
            }
         },
         "trackType" : "WebApollo/View/Track/DraggableNeatHTMLFeatures",
         "type" : "WebApollo/View/Track/DraggableNeatHTMLFeatures",
         "urlTemplate" : "tracks/NCBI_Annotation_Release_100_Gene/{refseq}/trackData.json"
         },
        {
         "category" : "TEST",
         "compress" : 0,
         "key" : "TEST_feature_by_type",
         "label" : "TEST_feature_by_type",
         "storeClass" : "JBrowse/Store/SeqFeature/NCList",
         "style" : {
            "className": "{type}",
            "description" : "product, note, description",
            "subfeatureClasses" : {
               "CDS" : "gnomon_CDS",
               "exon" : "container-100pct",
               "wholeCDS" : null
            }
         },
         "trackType" : "WebApollo/View/Track/DraggableNeatHTMLFeatures",
         "type" : "WebApollo/View/Track/DraggableNeatHTMLFeatures",
         "urlTemplate" : "tracks/NCBI_Annotation_Release_100_Gene/{refseq}/trackData.json"
      }
```
here is the JBrowse and plugins section of my apollo-config.groovy (commented out lines removed for clarity):
```
jbrowse {
   git {
       url= "https://github.com/GMOD/jbrowse"
       tag = "1.16.6-release"
       alwaysPull = true
       alwaysRecheck = true

        // Warning: We are still testing the performance of NeatFeatures plugins in combination with Apollo.
        // We advise caution if enabling these plugins with Apollo until this process is finalized.
   }
   plugins {
       WebApollo{
           included = true
       }
        NeatHTMLFeatures{
            included = true
        }
        NeatCanvasFeatures{
            included = true
        }
       RegexSequenceSearch{
           included = true
       }
       HideTrackLabels{
           included = true
       }
   }
}

```
Sample form my custom.css:
```
.jbrowse .gnomon_CDS,
.jbrowse .plus-gnomon_CDS,
.jbrowse .minus-gnomon_CDS {
    height: 12px;
    margin-top: 0px;
    background-color: #FFAA11;
}



```
I can share the full css too if that helps.
EDIT: adding a screenshot, and adding the config for the other track above.

<img width="1409" alt="Screen Shot 2020-01-29 at 11 30 17 AM" src="https://user-images.githubusercontent.com/421248/73376114-e6018200-428a-11ea-8181-477eda1cc4be.png">

**Comment by childers** [comment]
I have one update.  I removed the custom.css, and now the color by CDS function is working again.

Is there a better way/place to put in the CSS?

**Comment by nathandunn** [comment]
In the specific organism track directory is probably best. 

Nathan

**Comment by childers** [comment]
That is exactly where it is, right with the tracks.conf and trackList.json.

**Comment by nathandunn** [comment]
Apollo master should be fine .. that is 2.5.0, but double check.

Is it working the way you want?

Sometimes I have to play with the prefixes a bit using none, .jbrowse, or .apollo depending on the element.

**Comment by cmdcolin** [comment]
This isn't an exact answer but IIRC there are specific classes for the CDS frames that you could then override

**Comment by childers** [comment]
I mean the features take up the styles as defined by CSS, but turning "color by CDS" off and on has no effect on the feature tracks, though it still does work in the annotation panel.  When I remove my custom.css file, the "color by CDS" feature works everywhere as expected.

**Comment by nathandunn** [comment]
@childers  from https://genomearchitect.readthedocs.io/en/latest/Configure.html#setting-track-style-by-type:

- your `trackList.json` should point to your `custom.css` (which I think you already had)
- your custom.css should look like this probably (get rid of `.jbrowse` in favor of `.neat-UTR`:

```
.minus-gnomon_CDS  .neat-UTR,
.plus-gnomon_CDS  .neat-UTR,
.gnomon_CDS  .neat-UTR
{
    height: 12px;
    margin-top: 0px;
    background-color: #FFAA11;
}
```

Its possible that `.neat-UTR` is not correct.  You'll have to look at your computed styles to make sure. 

> After some discussion with @cmdcolin I opted to not use the !important flag in my css, to not force overrides.

While it is definitely best practice to avoid using `!important` as you aren't going to be re-inheriting those classes, I wouldn't let it keep you up at night if everything is working.

**Comment by childers** [comment]
I'm not sure I've stated the problem clearly.  We want the color by CDS feature to work for the evidence tracks, but when I add my custom css it breaks that feature.  Is that intentional?

**Comment by nathandunn** [comment]
>  Is that intentional?

Not this time ;)  

From my understanding, you are wanting to do color by type, which should now be supported in 2.5.0 by using CSS directly.

That should be done explicitly with CSS how I outlined [here](https://github.com/GMOD/Apollo/issues/2360#issuecomment-579965759), but its possible its simply not working, when you supply your own css. 

Do you have the full CSS file (let's try just a little bit) and trackList.json for the relevant portion and a screenshot? 

If a call would be easier, I would be open to that as well (email me).

**Comment by cmdcolin** [comment]
Alternatively, is the instance online we can see?

**Comment by childers** [comment]
@nathandunn coloring different features by type in one track works, but breaks the color by CDS feature, which we heavily promote to our annotators. That's the issue. If I specify my feature styles using CSS through any of the mentions I've worked out with you or Colin,  Color by CDS stops working for the evidence tracks.

@cmdcolin Unfortunately, I can't set up an instance for public viewing yet. That being said, I am happy to set you up with what you need.  These data are all openly available.

**Comment by nathandunn** [comment]
The likely solution/hack is to add the !important tag in a fork to webapollo_track_style.css so that cds frame class over-rides it:

```
.colorCds .cds-frame0 {
    background-color: #FF8080;
}

.colorCds .cds-frame1 {
    background-color: #80FF80;
}

.colorCds .cds-frame2 {
    background-color: #8080FF;
}

.colorCds .neg-cds-frame0 {
    background-color: #8080FF;
}

.colorCds .neg-cds-frame1 {
    background-color: #80FF80;
}

.colorCds .neg-cds-frame2 {
    background-color: #FF8080;
}
```

If anyone has a better idea

**Comment by nathandunn** [comment]
@cmdcolin  came up with a better solution (again) and @childers  will create a PR integrating it

**Comment by nathandunn** [comment]
fixed by #2362
